### PR TITLE
Sink: add in memory sink, slightly refactor validation/testing.

### PR DESCRIPTION
**Pull request by oschaaf**
- Adds an in-memory sink. Comes in handy in tests, and could serve
  a future role if we want to converge the way results are handled
  in both distributed and non-distributed tests.
- Slightly refactors validation & testing to make it easier to see
  what is specific to FileSinkImpl, making things a little more
  natural in case someone else wants to  add a new sink type in the
  future.
- Avoid MOCK_METHOD_N in the Sink mock as per recommendation.

Signed-off-by: Otto van der Schaaf <ovanders@redhat.com>

### CODE REVIEW

**Comment by mum4k** [comment]
@jiajunye please review and assign back to me once done.

**Comment by jiajunye** [comment]
LGTM

File: source/sink/sink_impl.cc [hunk 1]
```diff
@@ -36,12 +36,22 @@ absl::Status verifyCanBeUsedAsDirectoryName(absl::string_view s) {
   return absl::OkStatus();
 }
 
+absl::Status validateKey(absl::string_view key, const bool validate_as_directory_name) {
```
Comment (mum4k) [inline source/sink/sink_impl.cc:39]: (nit) Is it important for the bool argument to be const?

File: source/sink/sink_impl.cc [hunk 2]
```diff
@@ -104,4 +113,30 @@ FileSinkImpl::LoadExecutionResult(absl::string_view execution_id) const {
...
+InMemorySinkImpl::StoreExecutionResultPiece(const nighthawk::client::ExecutionResponse& response) {
+  absl::Status status = validateKey(response.execution_id(), false);
+  if (!status.ok()) {
+    return status;
+  }
+  auto iterator =
+      map_.insert({response.execution_id(), std::vector<nighthawk::client::ExecutionResponse>()})
+          .first;
+  iterator->second.push_back(response);
+  return absl::Status();
```
Comment (mum4k) [inline source/sink/sink_impl.cc:126]: I think the more conventional way of returning OK is:

```cc
return absl::OkStatus();
```

This one also works, but requires the reader to understand that an empty status defaults to OK.

https://abseil.io/blog/2020-091021-status

File: source/sink/sink_impl.cc [hunk 3]
```diff
@@ -104,4 +113,30 @@ FileSinkImpl::LoadExecutionResult(absl::string_view execution_id) const {
...
+InMemorySinkImpl::LoadExecutionResult(absl::string_view execution_id) const {
+  absl::Status status = validateKey(execution_id, false);
+  if (!status.ok()) {
+    return status;
+  }
+  auto iterator = map_.find(execution_id);
+  if (iterator != map_.end()) {
+    return (*iterator).second;
+  }
+  return absl::Status(absl::StatusCode::kNotFound, "Not found");
```
Comment (mum4k) [inline source/sink/sink_impl.cc:139]: (optional, fyi) Similarly to the comment above, a shorter version is:

```cc
return absl::NotFoundError("...");
```

Feel free to ignore, I know we have a few of these already in.

Comment (mum4k) [inline source/sink/sink_impl.cc:139]: Can we expand the error message, giving a better idea as to what wasn't found? Seeing an error message like this one might not be very helpful when debugging.

**Review by mum4k — CHANGES_REQUESTED** [review]

**Comment by oschaaf** [comment]
Thanks for the review @mum4k , comments addressed in 135e803b

**Review by mum4k — APPROVED** [review]

**Pull request merged**